File: bindings/php/tests/SecretboxTest.php
<?php

declare(strict_types=1);

use PHPUnit\Framework\TestCase;

/**
 * crypto_secretbox - three test categories per D-64/D-65: correctness (round trip - no official
 * vector exists for this construction, same posture as the Rust crate's own tests, D-51),
 * rejection (tamper/wrong key), misuse (wrong-length key, truncated input).
 */
final class SecretboxTest extends TestCase
{
    public function testRoundTripsSealOpen(): void
    {
        $key = dstu_core_secretbox_keygen();
        $sealed = dstu_core_secretbox_seal($key, 'a message worth protecting');
        $this->assertSame('a message worth protecting', dstu_core_secretbox_open($key, $sealed));
    }

    public function testHandlesAnEmptyMessage(): void
    {
        $key = dstu_core_secretbox_keygen();
        $sealed = dstu_core_secretbox_seal($key, '');
        $this->assertSame('', dstu_core_secretbox_open($key, $sealed));
    }

    public function testRejectsTamperedCiphertext(): void
    {
        $key = dstu_core_secretbox_keygen();
        $sealed = dstu_core_secretbox_seal($key, 'message');
        $sealed[strlen($sealed) - 1] = chr(ord($sealed[-1]) ^ 1);
        $this->expectException(DstuCoreException::class);
        dstu_core_secretbox_open($key, $sealed);
    }

    public function testRejectsTamperedNonce(): void
    {
        $key = dstu_core_secretbox_keygen();
        $sealed = dstu_core_secretbox_seal($key, 'message');
        $sealed[0] = chr(ord($sealed[0]) ^ 1);
        $this->expectException(DstuCoreException::class);
        dstu_core_secretbox_open($key, $sealed);
    }

    public function testRejectsWrongKey(): void
    {
        $key = dstu_core_secretbox_keygen();
        $otherKey = dstu_core_secretbox_keygen();
        $sealed = dstu_core_secretbox_seal($key, 'message');
        $this->expectException(DstuCoreException::class);
        dstu_core_secretbox_open($otherKey, $sealed);
    }

    public function testRejectsWrongLengthKey(): void
    {
        $this->expectException(\ValueError::class);
        dstu_core_secretbox_seal('too short', 'message');
    }

    public function testRejectsTruncatedSealedInput(): void
    {
        $key = dstu_core_secretbox_keygen();
        $this->expectException(DstuCoreException::class);
        dstu_core_secretbox_open($key, 'short');
    }
}

File: bindings/go/dstu/secretbox_test.go
package dstu

import (
	"bytes"
	"testing"
)

// SecretboxKey (crypto_secretbox) - correctness (round trip - no official vector exists for this
// construction, D-51), rejection (tamper/wrong key), misuse (wrong-length key, truncated input).

func TestSecretboxSealOpenRoundTrips(t *testing.T) {
	key, _ := GenerateSecretboxKey()
	defer key.Close()
	plaintext := []byte("a message worth protecting")
	sealed, err := key.Seal(plaintext)
	if err != nil {
		t.Fatal(err)
	}
	got, err := key.Open(sealed)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(got, plaintext) {
		t.Fatalf("got %q, want %q", got, plaintext)
	}
}

func TestSecretboxSealHandlesEmptyMessage(t *testing.T) {
	key, _ := GenerateSecretboxKey()
	defer key.Close()
	sealed, err := key.Seal(nil)
	if err != nil {
		t.Fatal(err)
	}
	got, err := key.Open(sealed)
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 0 {
		t.Fatalf("got %d bytes, want 0", len(got))
	}
}

func TestSecretboxTamperedCiphertextIsRejected(t *testing.T) {
	key, _ := GenerateSecretboxKey()
	defer key.Close()
	sealed, _ := key.Seal([]byte("message"))
	sealed[len(sealed)-1] ^= 1
	if _, err := key.Open(sealed); err == nil {
		t.Fatal("expected an error")
	}
}

func TestSecretboxTamperedNonceIsRejected(t *testing.T) {
	key, _ := GenerateSecretboxKey()
	defer key.Close()
	sealed, _ := key.Seal([]byte("message"))
	sealed[0] ^= 1
	if _, err := key.Open(sealed); err == nil {
		t.Fatal("expected an error")
	}
}

func TestSecretboxWrongKeyIsRejected(t *testing.T) {
	key, _ := GenerateSecretboxKey()
	defer key.Close()
	otherKey, _ := GenerateSecretboxKey()
	defer otherKey.Close()
	sealed, _ := key.Seal([]byte("message"))
	if _, err := otherKey.Open(sealed); err == nil {
		t.Fatal("expected an error")
	}
}

func TestSecretboxWrongLengthKeyIsRejected(t *testing.T) {
	if _, err := SecretboxKeyFromBytes(make([]byte, 10)); err == nil {
		t.Fatal("expected an error")
	}
}

func TestSecretboxTruncatedSealedInputIsRejected(t *testing.T) {
	key, _ := GenerateSecretboxKey()
	defer key.Close()
	if _, err := key.Open([]byte("short")); err == nil {
		t.Fatal("expected an error")
	}
}
